### DISCUSSION TITLE
Stack REM on ZNE

### DISCUSSION THREAD

**Discussion by hametar0u** [unitaryfund/mitiq#2251] [Q&A]
Hi,

I'm working through the tutorial on stacking ZNE on top of REM, and was trying to see if it was possible to change the order of the stacking. However, I'm running into the following error:

```
{
	"name": "TypeError",
	"message": "Result is not of type MeasurementResult.",
	"stack": "---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 69
     66 # NEW STUFF: stack REM on ZNE
     67 zne_rem_executor = rem.mitigate_executor(zne_executor, inverse_confusion_matrix=icm)
---> 69 zne_rem_result = zne_rem_executor(circuit)
     70 print(\"Mitigated value obtained with REM + ZNE:\", \"{:.5f}\".format(zne_rem_result.real))

File ~/Downloads/qrise-2024/.venv/lib/python3.11/site-packages/mitiq/rem/rem.py:99, in mitigate_executor.<locals>.new_executor(circuit)
     97 @wraps(executor)
     98 def new_executor(circuit: QPROGRAM) -> MeasurementResult:
---> 99     result = cast(MeasurementResult, executor_obj.run([circuit])[0])
    100     return result

File ~/Downloads/qrise-2024/.venv/lib/python3.11/site-packages/mitiq/executor/executor.py:272, in Executor.run(self, circuits, force_run_all, **kwargs)
    269     results_dict = dict(zip(collection.keys(), results))
    270     results = [results_dict[key] for key in hashable_circuits]
--> 272 return self._post_run(results)

File ~/Downloads/qrise-2024/.venv/lib/python3.11/site-packages/mitiq/rem/rem.py:85, in mitigate_executor.<locals>.post_run(self, results)
     81 def post_run(
     82     self: Executor,
     83     results: Sequence[MeasurementResult],
     84 ) -> Sequence[MeasurementResult]:
---> 85     return [
     86         mitigate_measurements(res, inverse_confusion_matrix)
     87         for res in results
     88     ]

File ~/Downloads/qrise-2024/.venv/lib/python3.11/site-packages/mitiq/rem/rem.py:86, in <listcomp>(.0)
     81 def post_run(
     82     self: Executor,
     83     results: Sequence[MeasurementResult],
     84 ) -> Sequence[MeasurementResult]:
     85     return [
---> 86         mitigate_measurements(res, inverse_confusion_matrix)
     87         for res in results
     88     ]

File ~/Downloads/qrise-2024/.venv/lib/python3.11/site-packages/mitiq/rem/inverse_confusion_matrix.py:185, in mitigate_measurements(noisy_result, inverse_confusion_matrix)
    173 \"\"\"Applies the inverse confusion matrix against the noisy measurement
    174 result and returns the adjusted measurements.
    175 
   (...)
    182     A mitigated MeasurementResult.
    183 \"\"\"
    184 if not isinstance(noisy_result, MeasurementResult):
--> 185     raise TypeError(\"Result is not of type MeasurementResult.\")
    187 num_qubits = noisy_result.nqubits
    188 required_shape = (2**num_qubits, 2**num_qubits)

TypeError: Result is not of type MeasurementResult."
}
```

Here is the code I am running:
```
import cirq
import numpy as np
from mitiq.benchmarks import generate_rb_circuits, ghz_circuits
from mitiq import MeasurementResult, Observable, PauliString, raw
from functools import partial

n_qubits = 2

circuit = generate_rb_circuits(2, 10)[0]

def execute(circuit: cirq.Circuit, noise_level: float = 0.002, p0: float = 0.05) -> MeasurementResult:
    """Execute a circuit with depolarizing noise of strength ``noise_level`` and readout errors ...
    """
    measurements = circuit[-1]
    circuit =  circuit[:-1]
    circuit = circuit.with_noise(cirq.depolarize(noise_level))
    circuit.append(cirq.bit_flip(p0).on_each(circuit.all_qubits()))
    circuit.append(measurements)

    simulator = cirq.DensityMatrixSimulator()

    result = simulator.run(circuit, repetitions=10000)
    bitstrings = np.column_stack(list(result.measurements.values()))
    return MeasurementResult(bitstrings)

obs = Observable(PauliString("ZI"), PauliString("IZ"))


# no QEM
noisy = raw.execute(circuit, execute, obs)
ideal = raw.execute(circuit, partial(execute, noise_level=0, p0=0), obs)
print("Ideal value:", "{:.5f}".format(ideal.real))
print("Unmitigated value:", "{:.5f}".format(noisy.real))



# REM
from mitiq import rem

p0 = p1 = 0.05
icm = rem.generate_inverse_confusion_matrix(2, p0, p1)
rem_executor = rem.mitigate_executor(execute, inverse_confusion_matrix=icm)

rem_result = obs.expectation(circuit, rem_executor)
print("Mitigated value obtained with REM:", "{:.5f}".format(rem_result.real))



# ZNE
from mitiq import zne

zne_executor = zne.mitigate_executor(execute, observable=obs, scale_noise=zne.scaling.folding.fold_global)
zne_result = zne_executor(circuit)
print("Mitigated value obtained with ZNE:", "{:.5f}".format(zne_result.real))



# Stack ZNE on top of REM
rem_zne_executor = zne.mitigate_executor(rem_executor, observable=obs, scale_noise=zne.scaling.folding.fold_global)

rem_zne_result = rem_zne_executor(circuit)
print("Mitigated value obtained with REM + ZNE:", "{:.5f}".format(rem_zne_result.real))



# NEW STUFF: stack REM on ZNE
zne_rem_executor = rem.mitigate_executor(zne_executor, inverse_confusion_matrix=icm)

zne_rem_result = zne_rem_executor(circuit)
print("Mitigated value obtained with REM + ZNE:", "{:.5f}".format(zne_rem_result.real))

```

<details>
<summary>
<code>pip freeze</code> output
</summary>
<pre>
anyio==4.3.0
appnope==0.1.4
asttokens==2.4.1
attrs==21.4.0
beautifulsoup4==4.12.3
bleach==6.1.0
cachetools==5.3.3
certifi==2024.2.2
charset-normalizer==3.3.2
cirq==1.3.0
cirq-aqt==1.3.0
cirq-core==1.3.0
cirq-ft==1.3.0
cirq-google==1.3.0
cirq-ionq==1.3.0
cirq-pasqal==1.3.0
cirq-rigetti==1.3.0
cirq-web==1.3.0
comm==0.2.2
contourpy==1.2.0
cycler==0.12.1
debugpy==1.8.1
decorator==5.1.1
defusedxml==0.7.1
Deprecated==1.2.14
duet==0.2.9
executing==2.0.1
fastjsonschema==2.19.1
fonttools==4.50.0
google-api-core==2.18.0
google-auth==2.29.0
googleapis-common-protos==1.63.0
grpcio==1.62.1
grpcio-status==1.62.1
h11==0.14.0
httpcore==0.16.3
httpx==0.23.3
idna==3.6
ipykernel==6.29.3
ipython==8.22.2
ipywidgets==8.1.2
iso8601==1.1.0
jedi==0.19.1
Jinja2==3.1.3
jsonschema==4.17.3
jupyter_client==8.6.1
jupyter_core==5.7.2
jupyterlab_pygments==0.3.0
jupyterlab_widgets==3.0.10
kiwisolver==1.4.5
lark==0.11.3
MarkupSafe==2.1.5
matplotlib==3.8.3
matplotlib-inline==0.1.6
mistune==3.0.2
mitiq==0.34.0
mpmath==1.3.0
msgpack==1.0.8
nbclient==0.10.0
nbconvert==7.16.3
nbformat==5.10.3
nest-asyncio==1.6.0
networkx==3.2.1
numpy==1.26.4
packaging==24.0
pandas==2.2.1
pandocfilters==1.5.1
parso==0.8.3
pexpect==4.9.0
pillow==10.2.0
platformdirs==4.2.0
prompt-toolkit==3.0.43
proto-plus==1.23.0
protobuf==4.25.3
psutil==5.9.8
ptyprocess==0.7.0
pure-eval==0.2.2
pyasn1==0.5.1
pyasn1-modules==0.3.0
pydantic==1.10.14
Pygments==2.17.2
PyJWT==2.8.0
pyparsing==3.1.2
pyquil==3.5.4
pyrsistent==0.20.0
python-dateutil==2.9.0.post0
python-rapidjson==1.16
pytz==2024.1
pyzmq==25.1.2
qcs-api-client==0.21.6
requests==2.31.0
retrying==1.3.4
rfc3339==6.2
rfc3986==1.5.0
rpcq==3.11.0
rsa==4.9
ruamel.yaml==0.18.6
ruamel.yaml.clib==0.2.8
scipy==1.11.4
seaborn==0.13.2
six==1.16.0
sniffio==1.3.1
sortedcontainers==2.4.0
soupsieve==2.5
stack-data==0.6.3
sympy==1.12
tabulate==0.9.0
tenacity==8.2.3
tinycss2==1.2.1
toml==0.10.2
tornado==6.4
tqdm==4.66.2
traitlets==5.14.2
types-Deprecated==1.2.9.20240311
types-python-dateutil==2.9.0.20240316
types-retry==0.9.9.4
typing_extensions==4.10.0
tzdata==2024.1
urllib3==2.2.1
wcwidth==0.2.13
webencodings==0.5.1
widgetsnbextension==4.0.10
wrapt==1.16.0
</pre>
</details>

**Comment by natestemen** [comment]
Awesome for giving this a shot!

The error you're running in to is due to the fact that the `execute` function you are using returns a [`mitiq.MeasurementResult`](https://mitiq.readthedocs.io/en/stable/apidoc.html#mitiq.typing.MeasurementResult) object which stores information about the raw measurements that were obtained from qubits (e.g. bitstrings). Zero-Noise Extrapolation is a technique that (unsurprisingly) has to perform extrapolation on the results that it obtains from the `execute` function. Since, in this case, they are not numbers, the extrapolation cannot occur.

The next follow on question would be _is it possible to use ZNE before REM?_ and the answer to that---as currently implemented in Mitiq---is no. [`execute_with_rem`](https://mitiq.readthedocs.io/en/stable/apidoc.html#mitiq.rem.rem.execute_with_rem) (and the other application methods of REM) all require the executor function to return bitstrings, which the output of ZNE can not return (without some serious modifications).

---

Small piece of advice for asking code related questions here (and other places, more generally): please try to find a minimal working example that demonstrates the issue/problem. The above issue actually arises purely from calling `mitiq.zne.mitigate_executor` on your `execute` function. REM is involved only insofar as the executor was defined in another context where it was being used.

No big deal, though! I appreciate the question more than anything :)

**Reply by hametar0u** [reply]
Hey @natestemen , thanks for the prompt reply!

I just had a few follow up questions; 

1. Do any other stacks work? I have gotten DDD + ZNE working, but REM + ZNE + DDD doesn't. Code example (follows from what I posted before):
```
# REM + ZNE + DDD
rem_zne_executor = zne.mitigate_executor(rem_executor, scale_noise=zne.scaling
.folding.fold_global, observable=obs)

rem_zne_ddd_executor = ddd.mitigate_executor(rem_zne_executor, observable=obs, rule=rule)

rem_zne_ddd_result = rem_zne_ddd_executor(circuit)
print("Mitigated value obtained with REM + DDD + ZNE:", "{:.5f}".format(rem_zne_ddd_result.real))
```
This throws:
```
ValueError: Cannot extract 2D measurements for repeated keys
```

2. I'm also trying to get QSE working on this circuit but I'm not sure why it is failing. Is it because QSE works in the same space as REM (ie. takes bitstrings and turns them into a number)? Code example (also follows from above):
```
# QSE
from mitiq import qse

Ms = [ # chat GPT generated stabilizer group
    "II",
    "XI",
    "ZI",
    "IX",
    "IZ"
]
check_operators = [
    PauliString(M, coeff=1, support=range(n_qubits)) for M in Ms
] # low key copy pasta-ed these so they may not even be the right choice
negative_Ms_as_pauliStrings = [
    PauliString(M, coeff=-1, support=range(n_qubits)) for M in Ms
] # low key copy pasta-ed these so they may not even be the right choice
code_hamiltonian = Observable(*negative_Ms_as_pauliStrings)

qse.execute_with_qse(
    circuit,
    execute,
    check_operators, # not sure what these are, need further reading
    code_hamiltonian,
    obs,
)
```

**Reply by natestemen** [reply]
> REM + ZNE + DDD

Have you tried this workflow without the `observable`? That might help simplify some of this as it shouldn't be required.

> QSE

The [demo](https://mitiq.readthedocs.io/en/stable/guide/qse-1-intro.html) uses an executor which returns a density matrix. Have you tried QSE with an executor that matches this signature?

**Reply by hametar0u** [reply]
Hi @natestemen,

Yes, I've tried without the `observable`. I'm getting this error: `ValueError: Circuit has no measurements to sample.`

Does this mean that nothing can be stacked on top of QSE? If I wanted to use the `mitigate_executor` function with it, then the circuit has to return a `MeasurementResult` type object which I don't think is the same as the density matrix. And if I were to put QSE at the back, then ZNE or something would somehow have to return a density matrix, and I don't think that's possible?

**Reply by natestemen** [reply]
> `ValueError: Circuit has no measurements to sample.`

Did adding measurements to the circuit fix this issue?

> QSE

I'm not actually totally sure, but happy to dig into it in another thread if you want to open another! Lets try to keep questions small and pointed so we can get to the bottom of them without getting sidetracked. It also makes the questions easier to peruse for others to find answers.

**Reply by hametar0u** [reply]
Hi @natestemen , apologies in advance if this is a rookie mistake as I am new to Cirq, but I have modified the circuit like so:

```
a, b = cirq.LineQubit.range(2)
obs = Observable(PauliString("ZZ"))
circuit = cirq.Circuit(
    cirq.H(a), cirq.CNOT(a, b), cirq.measure(a, b)
)
```

And I am getting the following error:
```
ValueError: Cannot extract 2D measurements for repeated keys
```

**Reply by natestemen** [reply]
That looks good for adding the measurements! So you can pass `circuit` into `rem_executor` to get a `MeasurementResult` object. If you then wish to pass this result on to ZNE you must convert it into a float. This can be handled by using an observable, as demonstrated in [this tutorial](https://mitiq.readthedocs.io/en/stable/examples/combine_rem_zne.html).

**Reply by hametar0u** [reply]
Hi @natestemen, I'm getting the same error. Here is the full code:

```
import cirq
import numpy as np
from mitiq.benchmarks import generate_rb_circuits
from mitiq import MeasurementResult, Observable, PauliString
from functools import partial

n_qubits = 2

# circuit = generate_rb_circuits(2, 10)[0]
a, b = cirq.LineQubit.range(2)
obs = Observable(PauliString("ZZ"))
circuit = cirq.Circuit(
    cirq.H(a), cirq.CNOT(a, b), cirq.measure(a, b)
)

def execute(circuit: cirq.Circuit, noise_level: float = 0.002, p0: float = 0.05) -> MeasurementResult:
    """Execute a circuit with depolarizing noise of strength ``noise_level`` and readout errors ...
    """
    measurements = circuit[-1]
    circuit =  circuit[:-1]
    circuit = circuit.with_noise(cirq.depolarize(noise_level))
    circuit.append(cirq.bit_flip(p0).on_each(circuit.all_qubits()))
    circuit.append(measurements)

    simulator = cirq.DensityMatrixSimulator()

    result = simulator.run(circuit, repetitions=3)
    bitstrings = np.column_stack(list(result.measurements.values()))
    return MeasurementResult(bitstrings)


# REM
from mitiq import rem

p0 = p1 = 0.05
icm = rem.generate_inverse_confusion_matrix(2, p0, p1)
rem_executor = rem.mitigate_executor(execute, inverse_confusion_matrix=icm)

# rem_result = obs.expectation(circuit, rem_executor)
# print("Mitigated value obtained with REM:", "{:.5f}".format(rem_result.real))


# Stack ZNE on top of REM
from mitiq import zne
rem_zne_executor = zne.mitigate_executor(rem_executor, observable=obs, scale_noise=zne.scaling.folding.fold_global)

rem_zne_result = rem_zne_executor(circuit)
print("Mitigated value obtained with REM + ZNE:", "{:.5f}".format(rem_zne_result.real))


# stack DDD on top of that
from mitiq import ddd
rule = ddd.rules.xx

rem_zne_ddd_executor = ddd.mitigate_executor(
    rem_zne_executor, rule=rule
)

rem_zne_ddd_result = rem_zne_ddd_executor(circuit)
print("Mitigated value obtained with REM + ZNE + DDD:", "{:.5f}".format(rem_zne_ddd_result.real))
```

**Reply by natestemen** [reply]
I'm not seeing where you convert the output of the `rem_executor` to a float. If you're not going to do that, then the circuit should not have measurements.

Please remember to post minimal working examples that reproduce the specific bug you are trying to resolve. Also make sure to let me know what you've tried and any progress you've made.

**Reply by hametar0u** [reply]
Apologies, I'm confused because I thought that `zne.mitigate_executor` takes in an executable, and not a single float. But for the purpose of answering your question, I tried doing
```
rem_result = obs.expectation(circuit, rem_executor)
```
But it's giving the same 2D measurement error thing.

**Reply by natestemen** [reply]
> I thought that zne.mitigate_executor takes in an executable

You're right! But it takes an executor with signature `circuit -> float`, rather than `circuit -> MeasurentResult`. When I said convert the output of `rem_executor` I meant define a new executor which takes the `MeasurementResult` data and returns some meaningful float attributed to such results. E.g. the probability that the state is in the ground state, or otherwise.
